File: MyDjango/static/gobang/js/computerplayer.js
var computerplayer = function(){
	this.computerstep = function (steps,cboard)
	{
		var color = ( steps%2 == 0?1:-1 );
		
		var beststep      = {i:-1,j:-1,score:-1};
		var enemybeststep = {i:-1,j:-1,score:-1};//对方的最佳落子点

		beststep =  py_step(steps,cboard);//here change to use python django compute 
		console.log(beststep);
		return beststep;
		
		for(var i =0;i < 15;i++)
		{
			for(var j = 0;j<15;j++)
			{				
				if(cboard[i][j][0] == 0)
				{
					var score      = getscore(cboard,i,j,color);
					var enemyscore = getscore(cboard,i,j,-color);
					if(beststep.score < score )
					{
						beststep.i = i;
						beststep.j = j;
						beststep.score = score;
						console.log(beststep);
					}
					if(enemybeststep.score < enemyscore )
					{
						enemybeststep.i = i;
						enemybeststep.j = j;
						enemybeststep.score = enemyscore;
						console.log(enemybeststep);
					}
				}
			}
		}
		if(beststep.score >= 10000)
			return beststep;
		else if(enemybeststep.score >= 10000)
		{
			return enemybeststep;
		}
		else if(beststep.score >= 1000)
			return beststep;
		else if(enemybeststep.score >= 1000)
		{
			return enemybeststep;
		}
		else
			return beststep;
		
	}

	function py_step(steps,cboard){
		console.log("in py_step");
		var strcheckerboard = "";
		for(var i=0;i<15;i++){
			for(var j=0;j<15;j++){
				strcheckerboard += checkerboard[i][j][0];
				if(i == 14 && j==14 )
					continue;
				else
					strcheckerboard += ",";
			}
		}
		console.log(checkerboard);
		console.log(strcheckerboard);
	
		var post_data = {
			"name": "testname",
			"steps":steps,
			"checkerboard":strcheckerboard
		};

		var rdata = {i:-1,j:-1,score:-1};
	
		$.ajax({
			url: "/gobang/py_step/",
			type: "POST",
			async: false,//使用同步的方式,true为异步方式.
			dataType:"json",//text
			data: post_data,
			success: function (data) {
				console.log("---------------successed---------")
				console.log(data);
				rdata.i = data["i"];
				rdata.j = data["j"];
				rdata.score = data["score"];
			},
			error: function(XMLHttpRequest, textStatus, errorThrown) 
			{
				alert(XMLHttpRequest.status);
				alert(XMLHttpRequest.readyState);
				alert(textStatus);
			},
			fail: function (err, status) {
				console.log(err)
			},
		});

		return rdata;
	}
	
	function getscore(cboard,i,j,color){
		return getRowScore(checkerboard,i,j,color)
			  +getColScore(checkerboard,i,j,color)
			  +getLeftSkewScore(checkerboard,i,j,color)
			  +getRightSkewScore(checkerboard,i,j,color);
	}
	
	var scoreArry = [0,1,10,100,1000,10000]
	
	function getRowScore(checkerboard,i,j,ncolor){
		var continuous = 1;
		for(var cj = j-1;cj>=0;cj--)
		{
			if(ncolor == checkerboard[i][cj][0])
			{
				continuous++;
			}
			else
			{
				break;
			}
		}
		for(var cj = j+1;cj<15;cj++)
		{
			if(ncolor == checkerboard[i][cj][0])
			{
				continuous++;
			}
			else
			{
				break;
			}
		}
		return scoreArry[continuous];
	}

	function getColScore(checkerboard,i,j,ncolor){
		var continuous = 1;
		for(var ci = i-1;ci>=0;ci--)
		{
			if(ncolor == checkerboard[ci][j][0])
			{
				continuous++;
			}
			else
			{
				break;
			}
		}
		for(var ci = i+1;ci<15;ci++)
		{
			if(ncolor == checkerboard[ci][j][0])
			{
				continuous++;
			}
			else
			{
				break;
			}
		}
		return scoreArry[continuous];
	}

	function getLeftSkewScore(checkerboard,i,j,ncolor){
		var continuous = 1;
		for(var ci = i-1,cj = j-1;ci>=0 && cj>=0;ci--,cj--)
		{
			if(ncolor == checkerboard[ci][cj][0])
			{
				continuous++;
			}
			else
			{
				break;
			}
		}
		for(var ci = i+1,cj = j+1;ci<15 && cj<15;ci++,cj++)
		{
			if(ncolor == checkerboard[ci][cj][0])
			{
				continuous++;
			}
			else
			{
				break;
			}
		}
		return scoreArry[continuous];
	}

	function getRightSkewScore(checkerboard,i,j,ncolor){
		var continuous = 1;
		for(var ci = i-1,cj = j+1;ci>=0 && cj<15;ci--,cj++)
		{
			if(ncolor == checkerboard[ci][cj][0])
			{
				continuous++;
			}
			else
			{
				break;
			}
		}
		for(var ci = i+1,cj = j-1;ci<15 && cj>=0;ci++,cj--)
		{
			if(ncolor == checkerboard[ci][cj][0])
			{
				continuous++;
			}
			else
			{
				break;
			}
		}
		return scoreArry[continuous];
	}

}

function deepcopy(obj) {
    var out = [],i = 0,len = obj.length;
    for (; i < len; i++) {
        if (obj[i] instanceof Array){
            out[i] = deepcopy(obj[i]);
        }
        else out[i] = obj[i];
    }
    return out;
}

var complayer = new computerplayer();

File: MyDjango/gobang/myModels/computerplayer.py
from ..fLogger import logger
import math
import json

scoreArry = [0,1,10,100,1000,10000]

class Computerplayter():
    def __init__(self):
        '''
           init  computer player
        '''
        self.checkboard =  [[0 for i in range(15)] for i in range(15)]
        

    def computestep(self, steps, strcheckboard):
        logger.info("computestep is called")
        #logger.info(strcheckboard)
        tcheckboard = strcheckboard.split(',')

        for i in range(len(tcheckboard)):
            self.checkboard[math.floor(i/15)][i%15] = int(tcheckboard[i])
        
        color = 1 if int(steps)%2==0 else -1
        beststep = {'i':-1,'j':-1,'score':-1}
        enemybeststep = {'i':-1,'j':-1,'score':-1}

        for i in range(15):
            for j in range(15):
                if self.checkboard[i][j] == 0:
                    score = self.getscore(i,j,color)
                    enemyscore = self.getscore(i,j,-color)

                    if beststep['score'] < score:
                        beststep["i"] = i
                        beststep["j"] = j
                        beststep["score"] = score

                    if enemybeststep['score'] < enemyscore:
                        enemybeststep["i"] = i
                        enemybeststep["j"] = j
                        enemybeststep["score"] = enemyscore
        
        if beststep["score"] >= 10000:
            return beststep
        elif enemybeststep["score"] >= 10000:
            return enemybeststep
        elif beststep["score"] >= 1000:
            return beststep
        elif enemybeststep["score"] >= 1000:
            return enemybeststep

        logger.info("in computerplayer.py beststep:"+json.dumps(beststep))
        return beststep

    def getscore(self, i, j, color):
        return self.getRowScore(i,j,color)\
			  +self.getColScore(i,j,color)\
			  +self.getLeftSkewScore(i,j,color)\
			  +self.getRightSkewScore(i,j,color)
    
    def getRowScore(self, i, j, color):
        continuous = 1
        cj = j-1
        while cj>=0:
            if self.checkboard[i][cj] == color:
                continuous += 1
                cj -= 1
            else:
                break
        cj = j+1
        while cj<15:
            if self.checkboard[i][cj] == color:
                continuous += 1
                cj += 1
            else:
                break
        return scoreArry[continuous]

    def getColScore(self, i, j, color):
        continuous = 1
        for ci in range(i-1,-1,-1):
            if self.checkboard[ci][j] == color:
                continuous += 1
            else:
                break
        for ci in range(i+1,14):
            if self.checkboard[ci][j] == color:
                continuous += 1
            else:
                break      
        return scoreArry[continuous]
    
    def getLeftSkewScore(self, i, j, color):
        continuous = 1
        ci = i-1
        cj = j-1
        while ci>=0 and cj>=0:
            if self.checkboard[ci][cj] == color:
                continuous += 1
                ci -= 1
                cj -= 1
            else:
                break
        ci = i+1
        cj = j+1
        while ci<15 and cj<15:
            if self.checkboard[ci][cj] == color:
                continuous += 1
                ci += 1
                cj += 1
            else:
                break
        return scoreArry[continuous]
    
    def getRightSkewScore(self, i, j, color):
        continuous = 1
        ci = i-1
        cj = j+1
        while ci>=0 and cj<15:
            if self.checkboard[ci][cj] == color:
                continuous += 1
                ci -= 1
                cj += 1
            else:
                break
        ci = i+1
        cj = j-1
        while ci<15 and cj>=0:
            if self.checkboard[ci][cj] == color:
                continuous += 1
                ci += 1
                cj -= 1
            else:
                break
        return scoreArry[continuous]




computerplayter = Computerplayter()